main.tf

```diff
@@ -7,10 +7,44 @@ locals {
   enable_bucket_logging = var.logging_bucket != ""
 }
 
+#
+# Enforce SSL/TLS on all transmitted objects
+# We do this by extending the custom_bucket_policy
+#
+
+data "aws_iam_policy_document" "enforcetls" {
+
+  source_json = var.custom_bucket_policy
+
+  statement {
+    sid    = "enforce-tls-requests-only"
+    effect = "Deny"
+    principals {
+      type        = "AWS"
+      identifiers = ["*"]
+    }
+    actions = ["s3:*"]
+    resources = [
+      "${aws_s3_bucket.private_bucket.arn}/*"
+    ]
+    condition {
+      test     = "Bool"
+      variable = "aws:SecureTransport"
+      values   = ["false"]
+    }
+  }
+}
+
+resource "aws_s3_bucket_policy" "bucket_policy" {
+  depends_on = [aws_s3_bucket_public_access_block.public_access_block]
+  bucket     = aws_s3_bucket.private_bucket.id
+  policy     = data.aws_iam_policy_document.enforcetls.json
+}
+
+
 resource "aws_s3_bucket" "private_bucket" {
   bucket = local.bucket_id
   acl    = "private"
-  policy = var.custom_bucket_policy
```

Comment:
Rather than creating a separate bucket policy resource you can render the JSON directly.

```hcl
policy = data. aws_iam_policy_document. enforce_tls.json
```

Reply:
There is a cyclic dependency issue with this since the name of the bucket is included in the policy. They will depend on each other.
```
Error: Cycle: module.s3_private_bucket.data.aws_iam_policy_document.enforce_tls, module.s3_private_bucket.aws_s3_bucket.private_bucket
```

The attachment was to circumvent that issue.

Reply:
Ah. I see what you mean now. I should have mentioned this in the first review. The trouble with switching to managing the bucket policy outside of the actual bucket is it creates brief window were the bucket will have no policy when running a terraform apply. This makes makes the upgrade path less than ideal and could be fixed by generating the s3 bucket ARN. So going from
```hcl
"${aws_s3_bucket.private_bucket.arn}/*"
```
into something like

```hcl
# to maintain govcloud compatibility
data "aws_partition" "current" {}
...
    resources = [
      "arn:${data.aws_partition.current.partition}:s3:::${local.bucket_id}/*"
    ]
...
```

Reply:
I ran out of time to further investigate what the upgrade path will look like. Anyway made necessary changes to bring policy withing the s3 bucket resource and built out the ARN. Also added partition variable for govcloud support.